File: packages/dprint-plugin-jsonc/src/tests/configurationTests.ts
import { expect } from "chai";
import { Configuration as GlobalConfiguration, ConfigurationDiagnostic } from "@dprint/types";
import { resolveConfiguration as resolveGlobalConfiguration, CliLoggingEnvironment } from "@dprint/core";
import { JsoncConfiguration, ResolvedJsoncConfiguration } from "../Configuration";
import { JsoncPlugin } from "../Plugin";

describe("configuration", () => {
    function doTest(
        config: JsoncConfiguration,
        expectedConfig: Partial<ResolvedJsoncConfiguration>,
        propertyFilter: (propName: keyof ResolvedJsoncConfiguration) => boolean,
        expectedDiagnostics: ConfigurationDiagnostic[] = [],
        globalConfig: Partial<GlobalConfiguration> = {},
    ) {
        const resolvedGlobalConfig = resolveGlobalConfiguration(globalConfig).config;
        const resolvedConfigResult = resolveConfiguration();
        const resolvedConfig = resolvedConfigResult.config;

        for (const propName in resolvedConfig) {
            if (!propertyFilter(propName as keyof ResolvedJsoncConfiguration))
                delete (resolvedConfig as any)[propName];
        }

        expect(resolvedConfig).to.deep.equal(expectedConfig);
        expect(resolvedConfigResult.diagnostics).to.deep.equal(expectedDiagnostics);

        function resolveConfiguration() {
            const plugin = new JsoncPlugin(config);
            try {
                plugin.initialize({
                    environment: new CliLoggingEnvironment(),
                    globalConfig: resolvedGlobalConfig,
                });
                return {
                    config: plugin.getConfiguration(),
                    diagnostics: plugin.getConfigurationDiagnostics(),
                };
            } finally {
                plugin.dispose();
            }
        }
    }

    describe("diagnostics", () => {
        it("should do a diagnostic when providing an incorrect number value", () => {
            doTest({ lineWidth: false as any as number }, {}, () => false, [{
                message: "Error parsing configuration value for 'lineWidth'. Message: invalid digit found in string",
                propertyName: "lineWidth",
            }]);
        });

        it("should do a diagnostic when providing an incorrect boolean value", () => {
            doTest({ "commentLine.forceSpaceAfterSlashes": 5 as any as boolean }, {}, () => false, [{
                message: "Error parsing configuration value for 'commentLine.forceSpaceAfterSlashes'. Message: provided string was not `true` or `false`",
                propertyName: "commentLine.forceSpaceAfterSlashes",
            }]);
        });

        it("should do a diagnostic when providing an excess property", () => {
            doTest({ asdf: 5 } as any, {}, () => false, [{
                message: "Unknown property in configuration: asdf",
                propertyName: "asdf",
            }]);
        });
    });

    describe(nameof<JsoncConfiguration>(c => c["commentLine.forceSpaceAfterSlashes"]), () => {
        function doSpecificTest(config: JsoncConfiguration, expectedConfig: Partial<ResolvedJsoncConfiguration>) {
            doTest(config, expectedConfig, prop => prop === "commentLine.forceSpaceAfterSlashes");
        }

        it("should get the default property", () => {
            doSpecificTest({}, { "commentLine.forceSpaceAfterSlashes": true });
        });

        it("should get the property when set", () => {
            doSpecificTest(
                { "commentLine.forceSpaceAfterSlashes": true },
                { "commentLine.forceSpaceAfterSlashes": true },
            );
        });
    });
});

File: packages/rust-core/src/configuration_tests.rs
use std::collections::HashMap;
use super::configuration::*;

#[test]
fn get_default_config_when_empty() {
    let config_result = resolve_global_config(HashMap::new());
    let config = config_result.config;
    assert_eq!(config_result.diagnostics.len(), 0);
    assert_eq!(config.line_width, None);
    assert_eq!(config.indent_width, None);
    assert_eq!(config.new_line_kind.is_none(), true);
    assert_eq!(config.use_tabs, None);
}

#[test]
fn get_values_when_filled() {
    let mut global_config = HashMap::new();
    global_config.insert(String::from("lineWidth"), String::from("80"));
    global_config.insert(String::from("indentWidth"), String::from("8"));
    global_config.insert(String::from("newLineKind"), String::from("crlf"));
    global_config.insert(String::from("useTabs"), String::from("true"));
    let config_result = resolve_global_config(global_config);
    let config = config_result.config;
    assert_eq!(config_result.diagnostics.len(), 0);
    assert_eq!(config.line_width, Some(80));
    assert_eq!(config.indent_width, Some(8));
    assert_eq!(config.new_line_kind == Some(NewLineKind::CarriageReturnLineFeed), true);
    assert_eq!(config.use_tabs, Some(true));
}

#[test]
fn get_diagnostic_for_invalid_enum_config() {
    let mut global_config = HashMap::new();
    global_config.insert(String::from("newLineKind"), String::from("something"));
    let diagnostics = resolve_global_config(global_config).diagnostics;
    assert_eq!(diagnostics.len(), 1);
    assert_eq!(diagnostics[0].message, "Error parsing configuration value for 'newLineKind'. Message: Found invalid value 'something'.");
    assert_eq!(diagnostics[0].property_name, "newLineKind");
}

#[test]
fn get_diagnostic_for_invalid_primitive() {
    let mut global_config = HashMap::new();
    global_config.insert(String::from("useTabs"), String::from("something"));
    let diagnostics = resolve_global_config(global_config).diagnostics;
    assert_eq!(diagnostics.len(), 1);
    assert_eq!(diagnostics[0].message, "Error parsing configuration value for 'useTabs'. Message: provided string was not `true` or `false`");
    assert_eq!(diagnostics[0].property_name, "useTabs");
}

#[test]
fn get_diagnostic_for_excess_property() {
    let mut global_config = HashMap::new();
    global_config.insert(String::from("something"), String::from("value"));
    let diagnostics = resolve_global_config(global_config).diagnostics;
    assert_eq!(diagnostics.len(), 1);
    assert_eq!(diagnostics[0].message, "Unknown property in configuration: something");
    assert_eq!(diagnostics[0].property_name, "something");
}
